From 16ea15c6f9503330038a432a73421db57772550b Mon Sep 17 00:00:00 2001
From: Dmitrii Shcherbakov <dmitrii.shcherbakov@canonical.com>
Date: Sat, 9 Mar 2019 02:51:18 +0300
Subject: [PATCH] py3: use binary mode for opening image files

If binary mode is not used when running with python3 glance image
uploads fail as files are opened with encoding='UTF-8' by default.

It results in the following confusing error message:

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfb in position 3:
invalid start byte

0xfb is a QCOW magic string and is invalid from UTF-8 standard
perspective. From the QCOW2 spec:

Byte  0 -  3:   magic
                QCOW magic string ("QFI\xfb")
---
 rally_openstack/services/image/glance_v1.py | 2 +-
 rally_openstack/services/image/glance_v2.py | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/rally_openstack/services/image/glance_v1.py b/rally_openstack/services/image/glance_v1.py
index bc5d6070..43f14072 100644
--- a/rally_openstack/services/image/glance_v1.py
+++ b/rally_openstack/services/image/glance_v1.py
@@ -52,7 +52,7 @@ class GlanceV1Service(service.Service, glance_common.GlanceMixin):
 
         try:
             if os.path.isfile(image_location):
-                kwargs["data"] = open(image_location)
+                kwargs["data"] = open(image_location, 'rb')
             else:
                 kwargs["copy_from"] = image_location
 
diff --git a/rally_openstack/services/image/glance_v2.py b/rally_openstack/services/image/glance_v2.py
index 070076ad..415fdb09 100644
--- a/rally_openstack/services/image/glance_v2.py
+++ b/rally_openstack/services/image/glance_v2.py
@@ -44,7 +44,7 @@ class GlanceV2Service(service.Service, glance_common.GlanceMixin):
         response = None
         try:
             if os.path.isfile(image_location):
-                image_data = open(image_location)
+                image_data = open(image_location, 'rb')
             else:
                 response = requests.get(image_location, stream=True)
                 image_data = response.raw
-- 
2.17.1

